### DISCUSSION TITLE
Training/Inference on GPU

### DISCUSSION THREAD

**Discussion by dr-moo** [umyelab/LabGym#105] [Q&A] [answered]
Hi!

First of all, I would like to thank you for this fantastic package; it is an absolute breakthrough for animal studies! 

I do, however, have some questions concerning GPU use. When I try installing LabGym 2.2 and training the initial detector, I get errors after the first epoch pointing towards inconsistency in CUDA/Cl. I know that Detectron2 is sensitive when it comes to torch/cuda versions, and I am struggling with setting it up right to utilize my GPU. Earlier versions of the torch specified by the Facebook team are incompatible with all the LabGym packages.

When I install the LabGym via pipx, I have no problems starting the training, but it runs on CPU, which is not optimal even when having 7950X. 

I would greatly appreciate any help so I could make it work on GPU.

If it is any easier, feel free to email me: oleksiy.guzhva@slu.se

Cheers!

**Comment by yujiahu415** [comment]
Hi, 
Thanks!
Is your system Windows or Linux? For Windows:

1. Install CUDA toolkit v11.8 (https://developer.nvidia.com/cuda-11-8-0-download-archive?target_os=Windows&target_arch=x86_64) and CUDNN library that supports 11.x (https://developer.nvidia.com/cudnn-downloads?target_os=Windows) (you need to register a free account to download cudnn). And add cuda/cudnn paths to the environment variables.

2. Install tensorflow v2.10.0 (pip install tensorflow==2.10.0) because after v2.10.0, tensorflow no longer support GPU on native Windows system (but the later version support Windows subsystem for Linux).

3. Install PyTorch v2.0.1+cu11.8 (pip install torch==2.0.1 torchvision==0.15.2 torchaudio==2.0.2 --index-url https://download.pytorch.org/whl/cu118)

If it's Linux system, still follow these steps and websites but choose Linux as the OS.

After these, you can import tensorflow and torch to see if GPU is detectable by these two libraries.

Let me know if these address your issue.

**Comment by dr-moo** [comment]
Hi Yujia and thank you so much for such a prompt response. I was getting slightly desperate yesterday when testing different scenarios. And the log from the Detectron2 team is not exactly up-to-date :-)

I am running the W11 system, but I have WSL2 installed as well.

I will give it a try and get back to you!

P.S. When installing via pipx, can the Detectron2 config be changed to direct the inference to GPU? Or if training the standalone Detectron2, can the checkpoint/weights be added to LabGym afterwards?

**Comment by yujiahu415** [comment]
No problem, Oleksiy! Let me know how it goes.

The installation via pipx is not suppose to change how LabGym operates. It's just an easier way for installation and user interface initialization. And it installs LabGym and its decencies in a separate environment I guess, so probably this is why you could initiate the training under pipx installation while failed to initiate under pip installation (global installation)---it's still the problem of CUDA version conflicts I guess.
@rohansatapathy @mvstru23 can you take a look at this? Thanks!

**Reply by dr-moo** [reply]
Currently getting the following errors when installing TensorFlow and Torch versions you mentioned in the intended conda environment with LabGym:

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
labgym 2.3.2 requires tensorflow>=2.15.0, but you have tensorflow 2.10.0 which is incompatible.
tensorflow-intel 2.15.0 requires keras<2.16,>=2.15.0, but you have keras 2.10.0 which is incompatible.
tensorflow-intel 2.15.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3, but you have protobuf 3.19.6 which is incompatible.
tensorflow-intel 2.15.0 requires tensorboard<2.16,>=2.15, but you have tensorboard 2.10.1 which is incompatible.
tensorflow-intel 2.15.0 requires tensorflow-estimator<2.16,>=2.15.0, but you have tensorflow-estimator 2.10.0 which is incompatible.

and 

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
labgym 2.3.2 requires tensorflow>=2.15.0, but you have tensorflow 2.10.0 which is incompatible.
labgym 2.3.2 requires torch>=2.1.2, but you have torch 2.0.1+cu118 which is incompatible.
labgym 2.3.2 requires torchvision>=0.16.2, but you have torchvision 0.15.2+cu118 which is incompatible.

It happens if i try to fix tthe dependencies  before trying to install the Detectron2 into my LabGym conda environment (which again, might not be surprising, considering the CUDA/Torch compatibility with Detectron2):

ERROR: Could not build wheels for detectron2, which is required to install pyproject.toml-based projects

If i just follow the installation manual for LabGym 2.2, it starts the GUI, but the training is on CPU, since it does not like CUDA 11.8 (all the path variables are correct, nvcc -V returns 11.8), and conda does not let me install pytorch with CUDA support. 

I did not try Detectron2 installation separatly, since the majority of examples use older pytorch versions which are not compatible with LabGym :-(

https://helloshreyas.com/how-to-install-detectron2-on-windows-machine

**Reply by yujiahu415** [reply]
These issues are basically version conflicts of CUDA / tensorflow / PyTorch. I have a Windows 11 system and also had the issues you had. I can provide information on how to troubleshoot them.

First, considering you are using Windows 11, I suggest you just switch to WSL. I managed to make everything working correctly on native Windows 11 except when training the Detectors, it gives a "WINERROR 10045" error or something like that stating that it cannot download the pretrained maskRCNN model from the remote host of Detectorn2 (this would not happen in Windows 10), which may requiring shut down firewall for the internet and I didn't proceed with that direction.

In WSL, after some configuration, it worked well except that the Categorizer (using tensorflow) on WSL is slower than that on native Windows, especially when you want to use hard disk paging for virtual memory when the system runs out of RAM. But it still works, and the Detector (using PyTorch) is actually faster on WSL than native Windows.

So here are the detailed steps (I may forget some steps but when you have the same issues I had it will remind me):

1. Follow this tutorial to install or update the Git and Microsoft VS C++ build tools: https://labgym.readthedocs.io/en/stable/installation/windows.html.
2. Reboot your system, and install CUDA Toolkit v11.8 and CUDNN library 8.6 on WSL2 (make sure it's WSL2 not WSL1, but Windows 11 should default it's WSL2). You can follow the steps listed on these websites: https://gist.github.com/MihailCosmin/affa6b1b71b43787e9228c25fe15aeba; https://gist.github.com/neouranos/51dd12d9471eb73b2348d373b605b45a; https://rachitsingh.com/notes/wsl-cuda/. And configure the CUDA paths / environmental variables on WSL2.
Important: DONT follow the instruction on the NVIDIA official website. It will get v12.3 installed, rather than v11.8, even if you select v11.8.
4. Upgrade pip, wheel, setuptools on WSL2. And then install LabGym through pip. You may encounter issues when building wheel of wxPython. If so, take a look at this: https://www.pixelstech.net/article/1599647177-Problem-and-Solution-for-Installing-wxPython-on-Ubuntu-20-04.
5. After LabGym is installed, install Detectron2: python3 -m pip install 'git+https://github.com/facebookresearch/detectron2.git'
6. After all is installed, you may need to configure the memory limit on WSL2 to make sure it can access to hard-disk virtual memory. To do so, you can create a ".wslconfig" file and put it on your WSL2 home folder. You can use text editor to edit the ".wslconfig" and configure "memory", "swap", and "autoMemoryReclaim", see "Example .wslconfig file" in: https://learn.microsoft.com/en-us/windows/wsl/wsl-config#wslconfig. Then configure the memory size as you wish.
7. tensorflow version typically is not a problem for WSL2, but Detectron2 may not be compatible with newer version of PyTorch. So I suggest you to install Install PyTorch v2.0.1+cu11.8 (pip install torch==2.0.1 torchvision==0.15.2 torchaudio==2.0.2 --index-url https://download.pytorch.org/whl/cu118).
8. Reboot your system.

Alternatively, if you still decide to go with native Windows, which is easier for installation steps, try to install LabGym==2.2.2 and tensorflow==2.10.0, also torch==2.0.1 torchvision==0.15.2 torchaudio==2.0.2 --index-url https://download.pytorch.org/whl/cu118, just ignore that warning of version conflict "
tensorflow-intel 2.15.0 requires keras<2.16,>=2.15.0, but you have keras 2.10.0 which is incompatible.
tensorflow-intel 2.15.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3, but you have protobuf 3.19.6 which is incompatible.
tensorflow-intel 2.15.0 requires tensorboard<2.16,>=2.15, but you have tensorboard 2.10.1 which is incompatible.
tensorflow-intel 2.15.0 requires tensorflow-estimator<2.16,>=2.15.0, but you have tensorflow-estimator 2.10.0 which is incompatible."
It should work (except for Detector training, which may require you to shut down firewall).

Let me know how this goes.

**Reply by rohansatapathy** [reply]
@dr-moo We have just released a new version of LabGym that fixes the compatibility issues with TensorFlow and PyTorch. If you upgrade to version 2.3.4 and follow the new [installation instructions](https://labgym.readthedocs.io/en/stable/installation/index.html), you should be able to get LabGym to work using your GPU. Please let us know if you run into any further issues with the installation, we're happy to help!

**Reply by dr-moo** [reply, marked as answer]
Hi again! I am updating you guys now after I have had time to sit and try different installation scenarios. 

I am running a workstation with W11 Beta Build 22635.3209. It's a new rig, so I am not yet done setting up the WSL with CUDA. I ran into too many issues that I had never encountered on my previous PCs, so I am about to give up, especially when my other dev environments run fine with separate conda/multiple local CUDA versions. 

But here is the story related to LabGym :-) It was maddening to install LabGym with Detectron2. And the main issue is Detectron itself. None of the instructions you guys gave me worked. It always failed with the last step when compiling the wheel for Detection. While it is known that Detectron does not play nice with Windows, I finally made it work (CUDA 11.8/cudnn 9.0).

1. Virtual conda environment with python=3.10
2. Within the environment: run pip install cython, and pip install Cmake
3. Add CMake to the Windows Environment PATH variables: It is very important to add CMake to the Environment PATH variables. To do this, download the latest release of CMake from https://perso.uclouvain.be/allan.barrea/opencv/building_tools.html and install it according to the instructions provided on the webpage.
4. Install LabGym: pip install labgym
5. Install the latest torch: pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118
6. From within the intended virtual environment: git clone https://github.com/facebookresearch/detectron2.git
7. Find the Detectron2 folder. For me, it got dumped into /Users/NAME/detectron2
8. In the detectron2 folder, open setup.py file and make sure you add this line "-DWITH_CUDA",  within the following snippet (right after: "-D__CUDA_NO_HALF2_OPERATORS__"):

if (torch.cuda.is_available() and ((CUDA_HOME is not None) or is_rocm_pytorch)) or os.getenv(
        "FORCE_CUDA", "0"
    ) == "1":
        extension = CUDAExtension
        sources += source_cuda

        if not is_rocm_pytorch:
            define_macros += [("WITH_CUDA", None)]
            extra_compile_args["nvcc"] = [
                "-O3",
                "-DCUDA_HAS_FP16=1",
                "-D__CUDA_NO_HALF_OPERATORS__",
                "-D__CUDA_NO_HALF_CONVERSIONS__",
                "-D__CUDA_NO_HALF2_OPERATORS__",
                 "-DWITH_CUDA",
            ]
        else:
            define_macros += [("WITH_HIP", None)]
            extra_compile_args["nvcc"] = []

10. Save setup.py
11.  Install Detectron2 with pip install -e detectron2
12.  Enjoy your LabGym experience!

**Reply by yujiahu415** [reply]
Thank you for these useful information! You're a pro!!

Yes, the most difficult part in LabGym installation was the installation of detectron2. And I saw you actually installed the latest PyTorch 2.2.1? My issues with detectron2 was that it was not compatible with any of PyTorch version that is greater than v2.0.1. I tried different versions and only the v2.0.1 did not give any errors when running detectron2 on my Windows 11 Home. So I had to tell users to downgrade PyTorch if their versions are greater than v2.0.1. So if you're using the Detector module (which is based on detectron2) in LabGym, please let me know how the latest PyTorch works with detectron2 on your system.

Another issue I encountered with the native Windows 11 was when training the Detectors, it gave a "WinError " that prevented the maskRCNN model to be downloaded from detectron2 host (I guess), which I currently don't have a clue how to solve it.

![1](https://github.com/umyelab/LabGym/assets/71684256/cc70a94e-b916-43cf-9238-af1a8f5845d5)

This error did not appear when on either Windows 10 or WSL OS.

**Reply by dr-moo** [reply]
Hi again, @yujiahu415! Now, I tried installing the LabGym on three different workstations running different Windows 11 Insider Builds (Canary, Dev, Beta) to test across varying levels of OS stability. Everything works with the latest Torch if you modify the local setup.py and use my tiny quick fix :-) LabGym is utilising the GPU in training, and I am also about to test the behavioural module. I am curious to see if everything works there and how fast the module is under native Windows.

**Reply by yujiahu415** [reply]
Thank you so much for the update!! @dr-moo you addressed one of my major concern! Let me know how it goes with the native Windows.
